cpp/src/toppra/parametrizer/const_accel.cpp

```diff
@@ -0,0 +1,129 @@
+#include <c++/7/bits/c++config.h>
+#include <cmath>
+#include <cstddef>
+#include <cstdlib>
+#include <stdexcept>
+#include <toppra/parametrizer.hpp>
+#include <toppra/parametrizer/const_accel.hpp>
+#include <toppra/toppra.hpp>
+
+namespace toppra {
+namespace parametrizer {
+
+ConstAccel::ConstAccel(GeometricPathPtr path, const Vector& gridpoints,
+                       const Vector& vsquared)
+    : Parametrizer(path, gridpoints, vsquared) {
+  process_internals();
+}
+
+void ConstAccel::process_internals() {
+  m_ts.resize(m_gridpoints.size());
+  m_us.resize(m_gridpoints.size() - 1);
+  m_ts[0] = 0;
+  for (std::size_t i = 0; i < m_gridpoints.size() - 1; i++) {
+    m_us[i] = 0.5 * (m_vs[i + 1] * m_vs[i + 1] - m_vs[i] * m_vs[i]) /
+              (m_gridpoints[i + 1] - m_gridpoints[i]);
+    m_ts[i + 1] =
+        m_ts[i] + 2 * (m_gridpoints[i + 1] - m_gridpoints[i]) / (m_vs[i + 1] + m_vs[i]);
+  }
+  TOPPRA_LOG_DEBUG("process_internals: ts=" << m_ts.transpose());
+  TOPPRA_LOG_DEBUG("process_internals: us=" << m_us.transpose());
+  TOPPRA_LOG_DEBUG("process_internals: vs=" << m_vs.transpose());
+}
+
+bool ConstAccel::evalParams(const Vector& ts, Vector& ss, Vector& vs,
+                            Vector& us) const {
+  TOPPRA_LOG_DEBUG("evalParams: ts=" << ts.transpose());
+  ss.resize(ts.size());
+  vs.resize(ts.size());
+  us.resize(ts.size());
+  int k_grid = 0;
+  for (std::size_t i = 0; i < ts.size(); i++) {
+    // find k_grid s.t m_gridpoints[k_grid] <= ss[i] < m_gridpoints[k_grid + 1]
+
+    // reset the search back to the first gridpoint, only do once
+    if (m_ts[k_grid] > ts[i]) {
+      k_grid = 0;
+    }
+
+    // increment k_grid until the condition is satisfied
+    while (k_grid < (m_ts.size() - 1)) {
+      if (m_ts[k_grid] <= ts[i] && ts[i] < m_ts[k_grid + 1]) {
+        break;
+      } else if (k_grid == (m_ts.size() - 2)) {
+        break;
+      } else {
+        k_grid++;
+      }
+    }
```

Comment:
FYI, this loop can be made logarithmic with the help of either [std::lower_bound](https://en.cppreference.com/w/cpp/algorithm/lower_bound) or [std::upper_bound](https://en.cppreference.com/w/cpp/algorithm/upper_bound).

If the number of grid point is high, this can make a substencial difference.

Reply:
Thanks for the suggestion. It makes perfect sense. Though I am not sure how to obtain begin/end iterators from a Eigen::Matrix. Do you have a suggestion?

Reply:
It is a vector so you should be able to do:
`std::lower_bound(m_ts.data(), m_ts.data()+m_ts.size(), value)`.

Reply:
https://stackoverflow.com/questions/27030550/foreach-loops-over-eigen-matrices

This suggests using x.data() and x.data() + x.size(). Tbh that looks a bit "unsafe" .

Reply:
I think it is explicitly written somewhere in the STL documentation that you can pass raw pointer in place of iterators. This is safe.